text/0000-location.md

```diff
@@ -0,0 +1,409 @@
+platform and may differ for other transaction execution platforms.
+
+Locations are managed by submitting transactions to Hyperledger Grid, which will
+process them with the Grid Location smart contract. The following transactions 
+are supported:
+- LocationCreate - create a Location and store it in state.
+- LocationUpdate - update (replace) the properties of a Location already in state.
+- LocationDelete - remove a Location from state.
+ 
+_*Disclaimer: LocationDelete is a potentially hazardous operation and needs to 
+be done with care._
+
+
+## Permissions
+[permissions]: #permissions
+
+Creation of GS1 locations is restricted to agents acting on behalf of the 
+organization that is associated with the GS1 company prefix encoded in the GLN. 
+An organization must have the GS1 company prefix in the �gs1_company_prefixes� 
+metadata field for its Pike organization. (Organization-level metadata will need
+to be implemented in Pike.) When a location is created, its owning organization 
+is stored with the location in an �owner� field.
+
+Updates to locations are restricted to agents acting on behalf of the 
+organization stored in the location�s owner field. Only property fields can be 
+updated. They are updated by providing a complete and updated list of the 
+properties, and overwrites the entire list of property values. Location_id, 
+location_namespace, and owner fields are immutable.
+
+Deletion of locations is restricted to agents acting on behalf of the 
+organization in the location�s owner field. A setting will turn off deletion 
+entirely, with the intent that it could be enabled/disabled by a Grid 
+administrator. Disabling delete is useful because external systems may have 
+references to these locations and deleting them could leave dangling references.
+
+
+## GLN Validation
+[gln validation]: #glnvalidation
+
+Creation of a reusable GLN validation function to programmatically express the 
+equation used to validate a GLN. It validates GLN format to avoid mistype errors
+similar to a credit card validation. It's implemented as an extensible function 
+such that further validation steps can be added, if needed. For details on the 
+equation see: [Check digit validation](https://www.gs1us.org/tools/check-digit-calculator "Check digit validation")
+
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## State
+[state]: #state
+
+### Location Representation
+
+The primary object stored in state is �Location�, which consists of a 
+location_id (GS1 GLN), an owner (org_id compatible with Pike), and a list of 
+property name/value pairs. The properties available are defined by the Grid 
+Location Schema Transaction Family and are restricted to the fields and rules of
+the GS1 Location Property Schema. Transactions which are responsible for setting
+location state values must ensure that the properties conform with the 
+requirements of the GS1 Location Property Schema.
+
```

Comment:
If we want to add in the protobuf message format, it would probably look like:

```
message Location { 
    enum LocationType { 
        UNSET_NAMESPACE = 0; 
        GS1 = 1; 
    }

    // What type of location is this (GS1)
    LocationType location_type = 1; 

    // location_id is the GLN for GS1-type locations
    string location_id = 2;

    // Who owns this location (pike organization id)
    string owner = 3;

    // Properties about this location
    repeated PropertyValue properties = 4; 
} 
```

Note that in the Product RFC, this is called ProductNamespace but the code is currently ProductType.

LocationType is technically important because even if we only ever implement GS1 locations, this is the technical mechanism by which we would add/extend location formats in the future.

Reply:
The RFC currently proposes a GS1 Attribute Name of "locationType" to store values such as "Ship From", "Bill To", etc. Will this cause conflict with the "LocationType location_type" used to indicate the GS1 namespace?

Reply:
@JessieZamzow, I have asked the GS1 US Standards team to take a look and see if they can offer any suggestions on this one.

Reply:
These will not conflict from a code/implementation perspective, though it will be potentially a bit confusing that the term has a different meaning depending on context. This might be good to add to the drawbacks section. Fixing this IMO would just require also updating Product so we are consistent, and if "ProductNamespace" and "LocationNamespace" aren't existing GS1 terms, we could decide to go back to that naming approach for these enums.

Reply:
GS1 US Standards team member here...

This is not a GS1 Standards question per se but our input could be helpful. In the GS1 System, there is a clear delineation between the GLN as a identification key for locations and attributes/ qualifiers which describe individual GLNs. Despite having clearly different roles within the protobuf message format, LocationType and location_type are close enough that users/implementers could be confused.

Reply:
To avoid confusion with the locationType attribute, I updated the RFC to reflect the "LocationNamespace" enum naming approach. Given the desire to drive consistency with Grid Product, I added a call-out within the Rationale and Alternatives section to update the Product codebase and created a related issue within the hyperledger / Grid repo. Please review and let me know if you agree with the changes. I don't have a strong code background so tactical suggestions are welcome. @vaporos If updates look okay then I believe you may resolve some below comments.

Reply:
Glad to see this naming issue get resolved. We opted for `product_namespce` after the smart contract was implemented using the `product_type` verbiage. The reason being that it's an overloaded term in the context of "products". Seems we reached a similar conclusion here, hooray for consistency! 

There was a PR that was opened back at the Hyperledger summit last year to make the code change from `product_type` to `product_namespace`. Though it seems the PR has gotten stale.: https://github.com/hyperledger/grid/pull/125